src/main/java/net/glowstone/util/PathFinder.java

```diff
@@ -0,0 +1,243 @@
+package net.glowstone.util;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.Set;
+
+import org.bukkit.util.Vector;
+
+/**
+ * General class for path finding.
+ * TODO: Implement other path finding algorithms (BFS, DFS, D*, potentially more)
+ */
+
+public class PathFinder {
+
+    private final Vector startNode;
+    private final Vector endNode;
+    private List<Vector> path;
+    private Map<Vector, Vector> fromPath;
+
+
+    /**
+     * Creates a new PathFinder object.
+     *
+     * @param startNode Starting position
+     * @param endNode Ending position
+     */
+    public PathFinder(final Vector startNode, final Vector endNode) {
+        // Save our start and end points
+        this.startNode = startNode;
+        this.endNode = endNode;
+        this.path = new ArrayList<Vector>();
+    }
+
+
+    /**
+     * Gets the start (initial) node.
+     *
+     * @return Vector object
+     */
+    public Vector getStartNode() {
+        return startNode;
+    }
+
+
+    /**
+     * Gets the end (goal) node.
+     *
+     * @return Vector object
+     */
+    public Vector getEndNode() {
+        return endNode;
+    }
+
+
+    /**
+     * Gets the path once it has been constructed.
+     *
+     * @note This method should only be called AFTER AStarSearch has been called
+     * @return List of Vector objects in order from start to end
+     */
+    public List<Vector> getPath() {
+        return path;
+    }
+
+
+    /**
+     * Tries to find a path between the start and end points. This method currently
+     * does not handle cases where a path could not potentially exist.
+     */
+    public void findPathAStar() {
+        Queue<Vector> frontier = new PriorityQueue<>();
```

Comment:
Hey we need a correct comparator!

Reply:
(The correct comparator should compare the current distances from the shortest path tree)

Reply:
I think I'll need to store the Vectors into a wrapper object that contains information on both the Vector and the distance and create the Queue based on that class.

Reply:
Not really. If the distance is updated, the element cannot be updated in the priority queue

Reply:
I have an idea of solving this. I have used this technique in programming competitions.
So we can't update the queue in realtime when the content of the node changes. How about we have a wrapper class for `Vector` that stores `Vector` and a `timeStamp`. When we need to update a `Vector`, we just create a new wrapper with new `Vector` and the newer `timeStamp` then add it to the queue. We also keep track of the newest `timeStamp` of all `Vector` in an array. In the `compareTo()` of the wrapper, we compare distance as usual. In the A* algorithm, whenever we take out a wrapper, we check if the `timeStamp` matches the `timeStamp` in the array. We discard wrappers with outdated `timeStamp`. Two must be equal for the wrapper to be accepted. 
You can see that we will have duplicated `Vector` in the queue, but each one of them will have different `timeStamp`. Only the one with the newest `timeStamp` will be accepted.
This solves the update problem without creating a custom priority queue. I made `timeStamp`an `int` start with 0. The A* keeps a master `timeStamp` and increases whenever an enqueue happens.